### PR TITLE
Python 3.13 support

### PR DESCRIPTION
**Pull request by guitargeek**
1. Sync with upstream cppyy, mostly to avoid warnings when building with Python 3.13.
2. Use GetOptional family of C Python API where necessary

A regression in ROOT 6.32 was unearthed in the process: #16852.

### CODE REVIEW

**Comment by guitargeek** [comment]
I opened a GitHub issue to document the underlying regression: https://github.com/root-project/root/issues/16852

This PR suggests to comment out for now the code that doesn't work anyway, for the sake of avoiding the crashes on Python 3.13.

The Fedora 41 is temporarily added to the CI to get test coverage for Python 3.13, but once the PR is shown to do what it should, the CI commit will be removed.

**Comment by guitargeek** [comment]
The failures on Fedora are all gone! Except for the expected one. Hence, I'll leave the Fedora 41 commit in this PR and restart the tests one more time.

**Review by vepadulano — APPROVED** [review]
Thank you! Feel free to merge once the CI is green

File: bindings/pyroot/cppyy/CPyCppyy/src/CPyCppyyModule.cxx [hunk 1]
```diff
@@ -214,23 +214,15 @@ static PyTypeObject PyDefault_t_Type = {
...
-#if PY_VERSION_HEX >= 0x30c00b1
-                                    {1},
-#else
-                                    1,
-#endif
-                                    &PyDefault_t_Type};
+PyObject _CPyCppyy_DefaultStruct = {
+    _PyObject_EXTRA_INIT
+    1, &PyDefault_t_Type
+};
```
Comment (hahnjo) [inline bindings/pyroot/cppyy/CPyCppyy/src/CPyCppyyModule.cxx:225]: This removes the changes done by @silverweed to avoid compiler warnings...

Comment (guitargeek) [reply bindings/pyroot/cppyy/CPyCppyy/src/CPyCppyyModule.cxx:225]: Ah! I have to re-introduce it then, also adding a patch file. Then I won't accidentally revert it anymore

**Pull request merged**